lib/Optimizer/Partition.cpp

```diff
@@ -0,0 +1,195 @@
+      continue;
+    }
+
+    // Start a new partition with this node.
+    auto *newF = F->getParent()->createFunction(
+        std::string(F->getName()) + "_part" + std::to_string(mapping.size()));
+    mapping.create(node, newF);
+  }
+
+  return mapping;
+}
+
+/// Given a function \p F and partitioning \p mapping, \return a FunctionGraph
+/// that contains appropriately-partitioned functions and their dependences.
+FunctionGraph doPartitioning(Function *F, NodeFunctionMap &mapping) {
+  FunctionGraph G(mapping.getFunctions());
+  llvm::DenseMap<Node *, Node *> currToNew;
+  auto *mod = F->getParent();
+
+  // Clone nodes into target partition.
+  for (auto &N : F->getNodes()) {
+    auto *clone = N.clone();
+    currToNew[&N] = clone;
+    mapping[&N]->addNode(clone);
+  }
+
+  // For any dependency that crosses a partition, add a variable and save
+  // node. Record the dependence in the function graph.
+  llvm::DenseMap<Node *, Variable *> variables;
+  for (auto *F : mapping.getFunctions()) {
+    for (auto &N : F->getNodes()) {
+      for (unsigned inp = 0, e = N.getNumInputs(); inp < e; inp++) {
+        auto input = N.getNthInput(inp);
+        if (isa<Variable>(input.getNode()))
+          continue;
+
+        auto *inputF = mapping[input.getNode()];
+        if (F == inputF)
+          continue;
+
+        // Add this dependence to the FunctionGraph.
+        G.add(F, inputF);
+
+        // If we've already created a variable for this dependence, use it.
+        auto it = variables.find(input.getNode());
+        if (it != variables.end()) {
+          N.setNthInput(inp, it->second);
+          continue;
+        }
+
+        // Create a new variable to represent this dependence.
+        auto *tmp = mod->createVariable(
+            input.getType(), std::string(input->getName()) + "_tmp",
+            VisibilityKind::Private, Variable::TrainKind::None);
+        inputF->createSave("tmp", input, tmp);
+        variables[input.getNode()] = tmp;
+        N.setNthInput(inp, tmp);
+      }
+    }
+  }
+
+  // Update links between nodes in the cloned functions.  Add variables (and
+  // save nodes) where a link crosses a partition boundary.
+  for (auto *F : mapping.getFunctions()) {
+    for (auto &N : F->getNodes()) {
+      for (unsigned inp = 0, e = N.getNumInputs(); inp < e; inp++) {
+        auto input = N.getNthInput(inp);
+
+        if (isa<Variable>(input.getNode()))
+          continue;
+
+        // Link this node to the clone of its input.
+        auto *clone = currToNew[input.getNode()];
+        N.setNthInput(inp, NodeValue(clone, input.getResNo()));
+      }
+    }
+  }
+
+  return G;
+}
+
+} // end namespace
+
+FunctionGraph::FunctionGraph(const FunctionList &functions)
+    : functions_(functions) {
+  for (auto *F : functions_) {
+    inputs_[F] = FunctionList();
+  }
+}
+
+FunctionGraph glow::partition(Function *F) {
+  NodeFunctionMap partitionMap = selectPartitions(F);
+  return doPartitioning(F, partitionMap);
```

Comment:
Maybe there should be a verifier here, e.g. to check that there are no cycles between Functions in the FunctionGraph returned from `doPartitioning()`?

Reply:
Great idea!  Done :-)